analyzer/src/main/kotlin/managers/Gradle.kt

```diff
@@ -120,13 +120,24 @@ class Gradle(
     private val dependencyHandler = GradleDependencyHandler(managerName, maven)
     private val graphBuilder = DependencyGraphBuilder(dependencyHandler)
 
+    // The path to the root project. In a single-project, just points to the project path. Is only null when not yet
+    // initialized.
+    private var rootProjectDir: File? = null
```

Comment:
Would it make sense to move the variable into `resolveDependencies()´ to avoid introducing a state?

Reply:
Unfortunately, we *need* to introduce state in order to "remember" the current root project path across `resolveDependencies()` calls.

Reply:
> Unfortunately, we _need_ to introduce state in order to "remember" the current root project path across `resolveDependencies()` calls.

I see, thanks.

I've researched a bit if there is any other way to determine the root project dir. After reading [1] I guess it may be possible to do something like the following (taken from [1]) but I haven't tried it:

`
ProjectConnection connection = GradleConnector.newConnector()
     .forProjectDirectory(getProjectDirPath().toFile())
     .connect();
project = connection.getModel(EclipseProject.class);
follow project.getParent() towards root
obtain the path
`

I'm not sure whether that above approach works at all and whether it would run itself into the issue you are trying to fix.

[1] https://www.programcreek.com/java-api-examples/?api=org.gradle.tooling.GradleConnector

Note: what I don't yet understand about this commit is what exact assumption about the ordering of the calls to `resolvedDependencies()` is introduced. And how it works for a repository with multiple root projects. Mind giving a hint?

Reply:
> what I don't yet understand about this commit is what exact assumption about the ordering of the calls to resolvedDependencies() is introduced.

IMO there's no assumption being *introduced*; but the change relies on the fact that `PackageManager.findManagedFiles()` returns files in depth-first order.

Reply:
> relies on the fact that `PackageManager.findManagedFiles()` returns files in depth-first order.

Thanks. IMO it'd be helpful to mention it in the comment above `private var rootProjectDir: File? = null`, as it needs quite a bit digging outside of the class to figure out where the order comes from.

Reply:
> `ProjectConnection connection = GradleConnector.newConnector()`

My fear is that we might run into a chicken-and-egg problem here: Connection to / evaluation / configuration of the project might fail if the `user.dir` property is not set as expected by the project, so we cannot use that approach to determine what `user.dir` needs to be set to.

Reply:
I still do not fully get whether the variable is always correctly reset. 

`
build.gradle
independent/build.gradle
child/gradle
`

If it processes in the order `build.gradle`, `independent/build.gradle`, `child/gradle`, would then the `rootProjectDir` be set to `independent` when processing `child`?

Reply:
> would then the `rootProjectDir` be set to `independent` when processing `child`?

No, the `rootProjectDir` stays at `./build.gradle`, because the `rootProjectDir` is only ever "reset" if a project is a root project, or an independent project / non-subproject.

Reply:
> > would then the `rootProjectDir` be set to `independent` when processing `child`?
> 
> No, the `rootProjectDir` stays at `./build.gradle`, because the `rootProjectDir` is only ever "reset" if a project is a root project, or an independent project / non-subproject.

My assumption was of course that `independent/build.gradle` is an independent / non-suproject. Have you considered this when writing above reply?

Reply:
> > No, the `rootProjectDir` stays at `./build.gradle`, because the `rootProjectDir` is only ever "reset" if a project is a root project, or an independent project / non-subproject.
> 
> My assumption was of course that `independent/build.gradle` is an independent / non-suproject. Have you considered this when writing above reply?

No.

Reply:
So, coming back to your original question, now that I got that the "independent" is not supposed to be a subproject of the root project:

> would then the `rootProjectDir` be set to `independent` when processing `child`?

No, but instead the `rootProjectDir` would be (wrongly) set to the directory containing the root "build.gradle" when processing "independent". So the project that gets the wrong `rootProjectDir` would be "independent", not "child". I don't currently see a feasible approach to improve this.

However, with the currently code "independent" would also get the wrong `rootProjectDir` (and all other projects, too). So from that perspective the current implementation still does the right thing for a sane project setup.

Reply:
> However, with the currently code "independent" would also get the wrong `rootProjectDir` (and all other projects, too). So from that perspective the current implementation still does the right thing for a sane project setup.

Do you mean, that the PR as-is fixes the issue in some case. In other cases it does not have any (negative) effect?

Reply:
Yes. It's not worse than before in some edge cases, but better than before in most cases, IMO.

Reply:
I'm ok if this was merged. I'd prefer to have some small comment in the code to highlight that this intentionally does not work in all cases.

Reply:
I just had added a `TODO`. Is its wording fine for you?

Reply:
Yes, thanks!